### PR TITLE
Simple tool to filter a BAM file to reduce size prior.

### PR DESCRIPTION
**Pull request by tfenne**
Useful when a lot of manipulations will be performed on the BAM file, e.g. repeatedly downsampling it.

### CODE REVIEW

**Comment by tfenne** [comment]
@nh13 This is not intended to be an all-conquering tool, just one that can filter a BAM down to a much smaller size quickly to make downstream processing more efficient if there is a lot of duplication and/or off-target.

File: src/main/java/com/fulcrumgenomics/bam/FilterBam.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
+package com.fulcrumgenomics.bam;
```
Comment (nh13) [inline src/main/java/com/fulcrumgenomics/bam/FilterBam.java:27]: license

File: src/main/java/com/fulcrumgenomics/bam/FilterBam.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    private final Log log = Log.getInstance(FilterBam.class);
+
+    // Default index creation to on
+    public FilterBam() { this.CREATE_INDEX = true; }
+
+    @Override
+    protected int doWork() {
+        IOUtil.assertFileIsReadable(INPUT);
+        IOUtil.assertFileIsWritable(OUTPUT);
+
```
Comment (nh13) [inline src/main/java/com/fulcrumgenomics/bam/FilterBam.java:89]: check if `INTERVALS` is readable if it is not `null` here?

File: src/main/java/com/fulcrumgenomics/bam/FilterBam.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+            if (REMOVE_DUPLICATES && rec.getDuplicateReadFlag()) continue;
+            if (REMOVE_UNMAPPED_READS && rec.getReadUnmappedFlag()) continue;
+            if (rec.getMappingQuality() < MIN_MAPQ) continue;
+            if (REMOVE_SECONDARY_ALIGNMENTS && !rec.getReadUnmappedFlag() && rec.getNotPrimaryAlignmentFlag()) continue;
+
+            ++kept;
+            out.addAlignment(rec);
+        }
+
+
```
Comment (nh13) [inline src/main/java/com/fulcrumgenomics/bam/FilterBam.java]: `\n`

File: src/main/java/com/fulcrumgenomics/bam/FilterBam.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    protected SAMRecordIterator buildInputIterator(final SamReader in, final File intervalListFile) {
+        if (intervalListFile == null) {
+            return in.iterator();
+        }
+        else {
+            IntervalList intervals = IntervalList.fromFile(intervalListFile).uniqued();
+            final SAMSequenceDictionary dict = intervals.getHeader().getSequenceDictionary();
+            final QueryInterval[] qs = intervals.getIntervals().stream()
+                    .map(i -> new QueryInterval(dict.getSequenceIndex(i.getContig()), i.getStart(), i.getEnd()))
+                    .toArray(QueryInterval[]::new);
```
Comment (nh13) [inline src/main/java/com/fulcrumgenomics/bam/FilterBam.java:129]: gorgeous

**Comment by nh13** [comment]
a few nitpicks, make the changes (or not), and you are ready to merge.

**Pull request merged**